### ISSUE TITLE
Issue with 'lm_logits' model

### ISSUE THREAD

**Issue by Anuj040** [kotikkonstantin/ru-autopunctuation#2]
Hey
Thanks a lot for making the work public, it is really helpful.
I am testing the code but there seems to be an issue with "lm_logits" model as the bert model does not return any logits. I wonder if there has been any change? 
I was thinking of working with other outputs of the model but none seems to make any sense. Any comments?
https://github.com/kotikkonstantin/ru-autopunctuation/blob/cdfbc5349f803434e88e889551f0bf596fa4d97f/model.py#L32

**Comment by kotikkonstantin** [comment]
Hi @Anuj040! Thanks for the question! What's the version of _'transformers'_ do you use? Is it like this https://github.com/kotikkonstantin/ru-autopunctuation/blob/master/requirements.txt#L3 ?

**Comment by Anuj040** [comment]
@kotikkonstantin As mentioned in the requirements.txt, 4.8.2

**Comment by kotikkonstantin** [comment]
@Anuj040 I've made the fix. You can check: it works or not?

**Comment by Anuj040** [comment]
It is working like a charm. Thanks a lot for looking into it. 

By the way, what strategies do you recommend for managing the imbalance in occurrence frequency of different "punctuation symbols" in the dataset?

I have tried one which seems to have some benefit over vanilla _cross-entropy loss_. Below is my example implementation in tensorflow. I basically weigh the positive side of _cross-entropy loss_ with ```max_frequency/(frequency of punctuation mark)```. Let me know your thoughts

```python
"""Custom loss to calculate weighted cross entropy loss"""

import tensorflow as tf


class WeightedCategoricalCrossentropy(tf.keras.losses.Loss):
    """Custom cross entropy loss to weigh the loss values depending on the class freq.

    Args:
        weights ([type]): weights for different classes
        from_logits (bool, optional): [description]. Defaults to False.
        label_smoothing (float, optional): Smoothen the labels to avoid overfitting. Defaults to 0.0.
        reduction ([type], optional): Defaults to tf.keras.losses.Reduction.SUM_OVER_BATCH_SIZE.
        axis (int, optional): [description]. Defaults to -1.
        name (str, optional): [description]. Defaults to "weighted_cce".
    """

    def __init__(
        self,
        weights,
        from_logits: bool = False,
        label_smoothing: float = 0.0,
        reduction=tf.keras.losses.Reduction.SUM_OVER_BATCH_SIZE,
        axis: int = -1,
        name: str = "weighted_cce",
    ):
        super().__init__()
        self.from_logits = from_logits
        self.label_smoothing = label_smoothing
        self.axis = axis
        self.weights = weights
        self.reduction = reduction

    def call(self, y_true: tf.Tensor, y_pred: tf.Tensor) -> tf.Tensor:
        """loss calculator"""
        if self.from_logits:
            logits = tf.nn.softmax(y_pred, axis=self.axis)
        else:
            logits = y_pred

        if self.label_smoothing > 0.0:
            y_true = self._smooth_labels(y_true)

        loss = self.weights * y_true * tf.math.log(logits) + (1 - y_true) * tf.math.log(
            1 - logits
        )
        return -tf.reduce_sum(loss) / tf.cast(tf.shape(y_true)[0], loss.dtype)

    def _smooth_labels(self, labels: tf.Tensor) -> tf.Tensor:
        """Smoothen the truth values. Helpful for preventing overfitting"""

        # smooth the labels
        labels *= 1.0 - self.label_smoothing
        labels += self.label_smoothing / tf.cast(tf.shape(labels)[-1], labels.dtype)

        # returned the smoothed labels
        return labels

**Comment by kotikkonstantin** [comment]
@Anuj040  Thank you very much for the question, idea, and script! It looks very informative!

Honestly, I didn't use any strategies to control class imbalance in this project. But I'd recommend approaching with combining the following three strategies:

- weighting objects of different classes (e.g., as you've suggested)

- sampling

- another loss (e.g., [Focal Loss](https://arxiv.org/pdf/1708.02002v2.pdf))

It will be better to try different ways to realize what's better for your task and domain if there is such an opportunity.
